crates/uv/src/commands/python/install.rs

```diff
@@ -193,35 +199,76 @@ pub(crate) async fn install(
 
     // Find requests that are already satisfied
     let mut changelog = Changelog::default();
-    let (satisfied, unsatisfied): (Vec<_>, Vec<_>) = requests.iter().partition_map(|request| {
-        if let Some(installation) = existing_installations
-            .iter()
-            .find(|installation| request.matches_installation(installation))
-        {
-            changelog.existing.insert(installation.key().clone());
-            if reinstall {
-                debug!(
-                    "Ignoring match `{}` for request `{}` due to `--reinstall` flag",
-                    installation.key().green(),
-                    request.cyan()
-                );
+    let (satisfied, unsatisfied): (Vec<_>, Vec<_>) = if reinstall {
+        // In the reinstall case, we want to iterate over all matching installations instead of
+        // stopping at the first match.
 
-                Either::Right(request)
-            } else {
+        let mut unsatisfied: Vec<Cow<InstallRequest>> =
+            Vec::with_capacity(existing_installations.len() + requests.len());
+
+        for request in &requests {
+            if existing_installations.is_empty() {
+                debug!("No installation found for request `{}`", request.cyan());
+                unsatisfied.push(Cow::Borrowed(request));
+            }
+
+            for installation in existing_installations
+                .iter()
+                .filter(|installation| request.matches_installation(installation))
+            {
+                changelog.existing.insert(installation.key().clone());
+                if matches!(&request.request, &PythonRequest::Any) {
+                    // Construct a install request matching the existing installation
+                    match InstallRequest::new(PythonRequest::Key(installation.into())) {
+                        Ok(request) => {
+                            debug!("Will reinstall `{}`", installation.key().green());
+                            unsatisfied.push(Cow::Owned(request));
+                        }
+                        Err(err) => {
+                            // This shouldn't really happen, but maybe a new version of uv dropped
+                            // support for a key we previously supported
+                            warn_user!(
+                                    "Failed to create reinstall request for existing installation `{}`: {err}",
+                                    installation.key().green()
+                                );
+                        }
+                    }
+                } else {
+                    // TODO(zanieb): This isn't really right! But we need `--upgrade` or similar
+                    // to handle this case correctly without causing a breaking change.
```

Comment:
Wait, what is this case?

Reply:
It's the one in this test case https://github.com/astral-sh/uv/pull/11072/files#diff-f499c776e1d8cc5e55d7620786e32e8732b675abd98e246c0971130f5de9ed50R157-R158

Reply:
We tell people to upgrade to the latest patch with `uv python install 3.13 --reinstall` so we can't change that behavior yet. Ideally that would _not_ change your patch version and you'd use `uv python install 3.13 --upgrade` to do that instead.

Reply:
Ok, so the `uv python install 3.13 --reinstall` behavior is unchanged, right? And we'll uninstall the existing `3.13`?

Reply:
Yep exactly. It's only if you do `uv python install any --reinstall` or `uv python install --reinstall` (and don't have a `.python-version` file) that this new behavior takes place.

Reply:
Yep exactly. It's only if you do `uv python install any --reinstall` or `uv python install --reinstall` (and don't have a `.python-version` file) that this new behavior takes place.